gapic/src/main/com/google/gapid/views/TracerDialog.java

```diff
@@ -903,7 +920,7 @@ private void downloadAndInstallAngle(Theme theme, Settings settings) {
 
         if (InstallAngleDialog
             .showDialogAndInstallApk(getShell(), models, theme, dev.device, url) == Window.OK) {
-          newAngleAvailableMessage.setVisible(false);
+          deviceLoader.notifyListeners(SWT.MouseDown, new Event());
```

Comment:
Hey John, I think the race condition sneaks in here. 
It's basically because we allow users to start tracing before device got actually refreshed. I could reproduce the gapir problem if I am trying to update ANGLE and I click "start tracing" very fast right after I close the install window.

By the current approach, the timeline looks like: 1. ANGLE got installed -> 2. user close ANGLE installing window -> 3. at this line, listener notifies `deviceLoader` to run -> 4. idle wait of 300ms(see Line365 of this file) -> 5. actually run `refreshDevices` runnable which also takes quite some time. If the user start tracing before step 5 finishes, the problem will occur. 

I think to solve it, we should mandate the `refreshDevices` to be the last step of the ANGLE installation process, so that everything is done before users closing the ANGLE installing window. To do that, I think we should basically modify `InstallAngleDialog` class: adding one more status, and pass in the `refreshDevices` runnable to be executed as the final step.

Reply:
Hi Stella, thank you for the review!

Actually after installing ANGLE the user is not able to start tracing before the device got refreshed, because the OK button will only be enabled if it detects ANGLE in the currently selected device, which can only happen after the refresh is complete. Also the trace would fail with "ANGLE not detected" if it would be run with the old device handle.

There must be another copy of the old device handle somewhere, which is used for opening the trace. This fails even when you don't take a new trace. It is enough to open an existing trace after installing ANGLE and cancelling the trace dialog.

Reply:
I was not able to solve the issue and was even able to reproduce it with enabled logging. Considering [Pascal's comment](https://github.com/google/agi/pull/799#discussion_r640063420) on Hughes identical solution I have reverted it and force the user instead after installing ANGLE to close and restart AGI.